### PR TITLE
Topological search ordering

### PR DESCRIPTION
**Pull request by tom-wa**
@markus2330 
not really sure how to implement actual tests for that part so for now there are just debug outputs for proof.
can you take a look and tell me if the way it works is fine with you ? i'll reformat it and remove the outputs afterwards

### CODE REVIEW

**Comment by markus2330** [comment]
Thank you for the update!

Testcases: simply do succeed_if for the correct key that should be there. With given ordering, the result should be deterministic anyway.

Do not forget to add docu, its still completely missing.

File: src/libs/meta/meta.c [hunk 1]
```diff
@@ -1003,33 +1003,62 @@ int elektraSortTopology (KeySet * ks, Key ** array)
...
+			keySetMeta (cur, "order", keyBaseName (orderCounter));
+			elektraArrayIncName (orderCounter);
+			ksAppendKey (done, keyDup (cur));
+			ksDel (deps);
+			continue;
+
+		    }
+		    else
+		    {
+			goto TopSearchElseBranch;
```
Comment (markus2330) [inline src/libs/meta/meta.c]: This control-flow needs some improvement.

File: src/libs/meta/meta.c [hunk 2]
```diff
@@ -931,6 +931,15 @@ void elektraMetaArrayAdd (Key * key, const char * metaName, const char * value)
 	keySetMeta (key, metaName, keyBaseName (arrayKey));
 	keyDel (arrayKey);
 }
+/*
+ * creates a KeySet from a MetaKey array.
```
Comment (markus2330) [inline src/libs/meta/meta.c]: Please use doxygen formatting.

File: src/libs/meta/meta.c [hunk 3]
```diff
@@ -989,6 +1002,14 @@ static int topCmpOrder (const void * a, const void * b)
 	return strcmp (keyString (kam), keyString (kbm));
 }
 
+/*
+ * does topological sorting on "ks" using the dependencies defined by the "dep/#" metakeys
```
Comment (markus2330) [inline src/libs/meta/meta.c]: How do the dep metakeys work? What about transitivity, duplicated deps,...

File: src/libs/meta/meta.c [hunk 4]
```diff
@@ -989,6 +1002,14 @@ static int topCmpOrder (const void * a, const void * b)
...
+ * if "order" metakeys are defined for the keys the algorithm tries to resolves them by that
+ * order.
+ * the sorted keys are stored in "array" in topological order
+ * returns 0 for cycles
```
Comment (markus2330) [inline src/libs/meta/meta.c]:
```
@retval -1 on wrong usage (what happens with array then?)
@retval 0 on cycles
@retval 1 on success
```

File: tests/ctest/test_meta.c [hunk 5]
```diff
@@ -300,12 +301,40 @@ static void checkTopArray (Key ** array, unsigned int size)
 	}
 	ksDel (done);
 }
+
+static void checkTopOrder1 (Key ** array)
+{
+	succeed_if (!strcmp (keyName (array[0]), "/d"), "Failed");
```
Comment (markus2330) [inline tests/ctest/test_meta.c]: defining your own macro for such cases is useful:
e.g. `succeed_if_top (0, "/d")`

File: src/libs/meta/meta.c (hunk 3 above)
Comment (tom-wa) [inline src/libs/meta/meta.c]: solving transitive dependencies is pretty much the core functionality of every topological sort algorithm, not sure what to say about that part.

Comment (markus2330) [inline src/libs/meta/meta.c]: That the depencencies are defined by the name? Ideally, you should write about how to use it and the specifics someone should know.

**Comment by markus2330** [comment]
Please ping if ready!

**Comment by tom-wa** [comment]
@markus2330 done

**Comment by markus2330** [comment]
Thank you!

**Pull request merged**